src/app/clusters/ota-requestor/BDXDownloader.cpp

```diff
@@ -18,101 +19,200 @@
+        VerifyOrReturn(err == CHIP_NO_ERROR, ChipLogError(BDX, "HandleBDXEvent: %" CHIP_ERROR_FORMAT, err.Format()));
+    } while (outEvent.EventType != TransferSession::OutputEventType::kNone);
+}
+
+CHIP_ERROR BDXDownloader::HandleBdxEvent(const chip::bdx::TransferSession::OutputEvent & outEvent)
+{
+    VerifyOrReturnError(mImageProcessor != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    switch (outEvent.EventType)
+    {
+    case TransferSession::OutputEventType::kNone:
+        break;
+    case TransferSession::OutputEventType::kAcceptReceived:
+        ReturnErrorOnFailure(mBdxTransfer.PrepareBlockQuery());
+        // TODO: need to check ReceiveAccept parameters
+        break;
+    case TransferSession::OutputEventType::kMsgToSend: {
+        VerifyOrReturnError(mMsgDelegate != nullptr, CHIP_ERROR_INCORRECT_STATE);
+        ReturnErrorOnFailure(mMsgDelegate->SendMessage(outEvent));
+        if (outEvent.msgTypeData.HasMessageType(chip::bdx::MessageType::BlockAckEOF))
         {
-            sendFlags.Set(chip::Messaging::SendMessageFlags::kExpectResponse);
+            // BDX transfer is not complete until BlockAckEOF has been sent
+            mState = State::kComplete;
+
+            // TODO: how/when to reset the BDXDownloader to be ready to handle another download
         }
-        err = mExchangeCtx->SendMessage(event.msgTypeData.ProtocolId, event.msgTypeData.MessageType, std::move(event.MsgData),
-                                        sendFlags);
-        VerifyOrReturn(err == CHIP_NO_ERROR, ChipLogError(BDX, "SendMessage failed: %" CHIP_ERROR_FORMAT, err.Format()));
         break;
     }
-    case TransferSession::OutputEventType::kAcceptReceived:
-        VerifyOrReturn(CHIP_NO_ERROR == mTransfer.PrepareBlockQuery(), ChipLogError(BDX, "PrepareBlockQuery failed"));
-        break;
     case TransferSession::OutputEventType::kBlockReceived: {
-        ChipLogDetail(BDX, "Got block length %zu", event.blockdata.Length);
+        chip::ByteSpan blockData(outEvent.blockdata.Data, outEvent.blockdata.Length);
+        ReturnErrorOnFailure(mImageProcessor->ProcessBlock(blockData));
 
-        // TODO: something more elegant than appending to a local file
-        // TODO: while convenient, we should not do a synchronous block write in our example application - this is bad practice
-        std::ofstream otaFile(outFilePath, std::ifstream::out | std::ifstream::ate | std::ifstream::app);
-        otaFile.write(reinterpret_cast<const char *>(event.blockdata.Data), static_cast<std::streamsize>(event.blockdata.Length));
-
-        if (event.blockdata.IsEof)
-        {
-            err = mTransfer.PrepareBlockAck();
-            VerifyOrReturn(err == CHIP_NO_ERROR, ChipLogError(BDX, "PrepareBlockAck failed: %" CHIP_ERROR_FORMAT, err.Format()));
-            mIsTransferComplete = true;
-        }
-        else
+        // TODO: this will cause problems if Finalize() is not guaranteed to do its work after ProcessBlock().
```

Comment:
What kind of problems and why would Finalize not be guaranteed to do its work?

Reply:
If `ProcessBlock()` schedules some async work but `Finalize()` does synchronous work, then it's possible `Finalize()` could execute before the last block has been written

Reply:
I see. So I'm not sure if Finalize should be called before ProcessBlock is done (e.g. OnBlockProcessed is called).

Perhaps one way to guarantee this is add a state to ImageProcessor to see if a block is being processed and if so, any calls to `Finalize` should result in an immediate error indicating "busy" or something?